src/backend/distributed/operations/shard_cleaner.c

```diff
@@ -769,12 +768,110 @@ TryLockRelationAndPlacementCleanup(Oid relationId, LOCKMODE lockmode)
 
 
 /*
- * TryDropShard tries to drop the given shard placement and returns
+ * TryDropResourceByCleanupRecordOutsideTransaction tries to drop the given resource
+ * and returns true on success.
+ */
+static bool
+TryDropResourceByCleanupRecordOutsideTransaction(CleanupRecord *record,
+												 char *nodeName,
+												 int nodePort)
+{
+	List *dropCommandList = NIL;
+
+	switch (record->objectType)
+	{
+		case CLEANUP_OBJECT_SHARD_PLACEMENT:
+		{
+			return TryDropShardOutsideTransaction(record->objectName,
+												  nodeName, nodePort);
+		}
+
+		case CLEANUP_OBJECT_SUBSCRIPTION:
+		{
+			StringInfo disableQuery = makeStringInfo();
+			appendStringInfo(disableQuery,
+							 "ALTER SUBSCRIPTION %s DISABLE",
```

Comment:
This is not an idempotent way of dropping the subscription. If the subscription was already dropped, this will fail. We already have functions to drop these resources in an idempotent way. I think we should reuse those (possibly with necessary changes).

Reply:
It seems you removed these functions in this PR, but I'm talking about DropSubscription, DropPublication, DropReplicationSlot, DropUser. I think it's better to keep those functions and simply call them here. It also keeps the code a bit separate.

Reply:
Thanks. It got better now